core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -0,0 +1,378 @@
+
+                // reroute now to remove any dead nodes (master may have stepped down when they left and didn't update the routing table)
+                RoutingAllocation.Result result = routingService.getAllocationService().reroute(currentState);
+                if (result.changed()) {
+                    currentState = ClusterState.builder(currentState).routingResult(result).build();
+                }
+
+                // add the incoming join requests
+                return super.execute(currentState);
+            }
+
+            @Override
+            public boolean runOnlyOnMaster() {
+                return false;
+            }
+
+            @Override
+            public void onFailure(String source, Throwable t) {
+                super.onFailure(source, t);
+                context.onFailure(t);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                super.clusterStateProcessed(source, oldState, newState);
+                context.onElectedAsMaster(newState);
+            }
+        });
+    }
+
+    private void processJoins(String reason) {
+        clusterService.submitStateUpdateTask("zen-disco-join(" + reason + ")", Priority.URGENT, new ProcessJoinsTask());
+    }
+
+
+    public interface Callback {
+        void onElectedAsMaster(ClusterState state);
+
+        void onFailure(Throwable t);
+    }
+
+    static abstract class ElectionContext implements Callback {
+        private final Callback callback;
+        private final int requiredMasterJoins;
+
+        /** set to true after enough joins have been seen and a cluster update task is submitted to become master */
+        final AtomicBoolean pendingSetAsMasterTask = new AtomicBoolean();
+        final AtomicBoolean closed = new AtomicBoolean();
+
+        ElectionContext(Callback callback, int requiredMasterJoins) {
+            this.callback = callback;
+            this.requiredMasterJoins = requiredMasterJoins;
+        }
+
+        abstract void onClose();
+
+        @Override
+        public void onElectedAsMaster(ClusterState state) {
+            assert pendingSetAsMasterTask.get() : "onElectedAsMaster called but pendingSetAsMasterTask is not set";
+            if (closed.compareAndSet(false, true)) {
+                try {
+                    onClose();
+                } finally {
+                    callback.onElectedAsMaster(state);
+                }
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            if (closed.compareAndSet(false, true)) {
+                try {
+                    onClose();
+                } finally {
+                    callback.onFailure(t);
+                }
+            }
+        }
+    }
+
+
+    class ProcessJoinsTask extends ProcessedClusterStateUpdateTask {
```

Comment:
maybe make this class abstract? I assume the idea to extract code from the checkPendingJoinsAndElectIfNeeded() method, is that it otherwise gets difficult to read?